gst-plugin/src/object.rs

```diff
@@ -1,506 +1,53 @@
-// Copyright (C) 2017 Sebastian Dröge <sebastian@centricular.com>
-//
-// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
-// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
-// option. This file may not be copied, modified, or distributed
-// except according to those terms.
-
-use std::any::TypeId;
-use std::collections::BTreeMap;
-use std::ffi::CString;
-use std::mem;
 use std::ptr;
-use std::sync::Mutex;
 use std::sync::atomic::AtomicBool;
 
-use glib_ffi;
-use gobject_ffi;
-
-use glib;
-use glib::translate::*;
+pub use gobject_base::object::*;
 
-use properties::*;
 
-pub trait ObjectImpl<T: ObjectType>: Send + Sync + 'static {
-    fn set_property(&self, _obj: &glib::Object, _id: u32, _value: &glib::Value) {
-        unimplemented!()
-    }
 
-    fn get_property(&self, _obj: &glib::Object, _id: u32) -> Result<glib::Value, ()> {
-        unimplemented!()
-    }
+#[repr(C)]
+pub struct InstanceStruct<T: ObjectType>
+{
+    _parent: T::GlibType,
+    _imp: ptr::NonNull<T::ImplType>,
 
-    fn notify(&self, obj: &glib::Object, name: &str) {
-        unsafe {
-            gobject_ffi::g_object_notify(obj.to_glib_none().0, name.to_glib_none().0);
-        }
-    }
+    _panicked: AtomicBool,
 }
 
-#[macro_export]
-macro_rules! box_object_impl(
-    ($name:ident) => {
-        impl<T: ObjectType> ObjectImpl<T> for Box<$name<T>> {
-            fn set_property(&self, obj: &glib::Object, id: u32, value: &glib::Value) {
-                let imp: &$name<T> = self.as_ref();
-                imp.set_property(obj, id, value);
-            }
-
-            fn get_property(&self, obj: &glib::Object, id: u32) -> Result<glib::Value, ()> {
-                let imp: &$name<T> = self.as_ref();
-                imp.get_property(obj, id)
-            }
-        }
-    };
-);
-
-pub trait ImplTypeStatic<T: ObjectType>: Send + Sync + 'static {
-    fn get_name(&self) -> &str;
-    fn new(&self, &T) -> T::ImplType;
-    fn class_init(&self, &mut ClassStruct<T>);
-    fn type_init(&self, _: &TypeInitToken, _type_: glib::Type) {}
+pub trait PanicPoison{
```

Comment:
Something like `ElementInstanceStructType` maybe? But that's quite long

Reply:
If we rename ```InstanceStruct``` to ```ObjectInstance```, we could use ```ElementInstance``` instead. However, then the mirroring to ClassStruct is less clear.

Using a trait here for the panic bool enabled the use of trait bounds. The compiler was happier about that than using the gst-plugin defined InstanceStruct everywhere. Still, the name bothers me a bit.

Perhaps it should just be called ```Panic```? In the end, ```Drop``` or ```Send``` are just as generic.

Reply:
The instance struct is fine now as-is as `ElementInstanceStruct`. This trait maybe just keep as `PanicPoison`, that's exactly what it is in the end